text/0079-cdk-v2.md

```diff
@@ -0,0 +1,339 @@
+a module for each service they use, even simple projects end up with dozens of
+direct CDK dependencies.
+
+## Experimental Modules
+
+In some cases, a CDK module can include both stable and experimental constructs.
+For example, a module like **@aws-cdk/aws-eks** includes stable CFN resources (L1s)
+but also a variety of experimental L2 classes.
+
+Our API stability guarantees and adherence to semantic versioning only applies
+to STABLE APIs and not to EXPERIMENTAL APIs.
+
+This situation is confusing to users. It is not uncommon for users to use
+experimental constructs without being aware and then be surprised by breaking
+changes in minor versions.
+
+The current approach makes it hard for us to specify the stability at the module
+level since the L1s are stable by definition, but all high-level constructs are
+still experimental.
+
+## Reusing constructs
+
+The AWS CDK demonstrates that leveraging general purpose programming languages
+to define complex configuration modules brings value to users. At the heart of
+the AWS CDK lies the "constructs programming model" which consists of a simple
+pattern that enables composition and synthesis.
+
+Recently, there were a few initiatives to build CDKs for other domains with
+similar characteristics. Examples are Kubernetes, Terraform and GitHub Actions.
+
+We wish to enable the creation of these "sister" CDK projects by extracting the
+constructs programming model into an independent library, which can be leveraged
+for any domain without taking a dependency on the AWS CDK.
+
+In order to enable composition of constructs across domains (e.g. Kubernetes app
+defined within an AWS CDK app), we will need to extract the `Construct` base
+class from the @aws-cdk/core module and into a separate module and AWS CDK users
+will need to take a dependency on this new module.
+
+# Design Summary
+
+To address the issues related to **dependencies**, we propose to package the
+entire AWS CDK as a single module, where each [sub]module will be represented
+through a namespace. This has the potential to address both issues related to
+dependencies: it will be possible to model this single AWS CDK module as a
+single peer dependency; and also users will need to only install a single module
+to use the entire AWS CDK. This will dramatically improve the consumption
+ergonomics.
+
+To improve **transparency about experimental modules**, we propose to move all
+experimental classes into a separate [sub]module with the `-experimental`
+postfix (e.g. `@aws-cdk/aws-eks-experimental`). This means that the non
+"-experimental" modules will always have stable APIs (including L1s) and will
+give us a way to graduate classes from the experimental module to the stable
+module. Users who use a module with the name "xxx-experimental" will have more
+awareness that this API may break in minor versions.
+
+To **enable the reuse of the constructs programming model** we will extract and
```

Comment:
I'm in favor of this. Do be aware that it will be added to the list of `peerDependencies` that has to be resolved by users. That is probably still fine, but keep this in mind.

Reply:
Yes, totally aware. Will make sure it is called out.